federation/pkg/federation-controller/service/servicecontroller.go

```diff
@@ -406,34 +406,33 @@ func (s *ServiceController) deleteClusterService(clusterName string, cachedServi
 
 func (s *ServiceController) ensureClusterService(cachedService *cachedService, clusterName string, service *v1.Service, client *kubeclientset.Clientset) error {
 	var err error
-	var needUpdate bool
 	for i := 0; i < clientRetryCount; i++ {
 		svc, err := client.Core().Services(service.Namespace).Get(service.Name)
 		if err == nil {
 			// service exists
 			glog.V(5).Infof("Found service %s/%s from cluster %s", service.Namespace, service.Name, clusterName)
-			//reserve immutable fields
-			service.Spec.ClusterIP = svc.Spec.ClusterIP
+			// we tend to update the new spec and
+			// current service spec(might have random cluster ip and node port) in the cluster identical
+
+			// cluster ip is not specified, update the auto assigned one to new service spec
+			if service.Spec.ClusterIP == "" {
+				service.Spec.ClusterIP = svc.Spec.ClusterIP
+			}
 
-			//reserve auto assigned field
 			for i, oldPort := range svc.Spec.Ports {
 				for _, port := range service.Spec.Ports {
+					// node port is not specified, update the auto assigned one to new service spec
 					if port.NodePort == 0 {
 						if !portEqualExcludeNodePort(&oldPort, &port) {
-							svc.Spec.Ports[i] = port
-							needUpdate = true
-						}
-					} else {
-						if !portEqualForLB(&oldPort, &port) {
-							svc.Spec.Ports[i] = port
-							needUpdate = true
+							service.Spec.Ports[i] = oldPort
 						}
 					}
 				}
 			}
 
-			if needUpdate {
-				// we only apply spec update
+			if util.ObjectMetaAndSpecEquivalent(svc, service) {
+				glog.V(4).Infof("Service %q in cluster %q does not need an update: cluster service is equivalent to federated service", svc, clusterName)
```

Comment:
nit: Output both the cluster service and the federated service in this log message.